text/0000-build-std/1-background.md

```diff
@@ -0,0 +1,309 @@
+it to the extern prelude. This could allow a path for crates like `alloc` or
+`test` to be provided without effecting the observable behaviour of the
+language.
+
+## Panic strategies
+[background-panic-strategies]: #panic-strategies
+
+Rust has the concept of a *panic handler*, which is a crate that is responsible
+for performing a panic. There are various panic handler crates on crates.io,
+such as [panic-abort] (which different from the `panic_abort` panic runtime!),
+[panic-halt], [panic-itm], and [panic-semihosting]. Panic handler crates define
+a function annotated with `#[panic_handler]`. There can only be one
+`#[panic_handler]` in the crate graph.
+
+`core` uses the panic handler to implement panics inserted by code generation
+(e.g. arithmetic overflow or out-of-bounds access) and the `core::panic!` macro
+immediately delegates to the panic handler crate.
+
+`std` is also a panic handler. `std`'s panic handler and `std::panic!` macro
+print panic information to stderr and delegate to a *panic runtime* to decide
+what to do next, determined by the *panic strategy*.
+
+There are two panic runtime crates in the standard library - `panic_unwind` and
+`panic_abort` - each with a corresponding panic strategy. Each target supported
+by rustc specifies a default panic strategy - either "unwind" or "abort" -
+though these are only relevant if `std`'s panic handler is used (i.e. the target
+isn't a `no_std` target or being used with a `no_std` crate).
+
+Rust's `-Cpanic` flag allows the user to choose the panic strategy, with the
+target's default as a fallback. If `-Cpanic=unwind` is provided then this
+doesn't guarantee that the unwind strategy is used, as the target may not
+support it.
+
+Both crates are compiled and shipped with the pre-built standard library for
+targets which support `std`. Some targets have a pre-built standard library with
+only the `core` and `alloc` crates, such as the `x86_64-unknown-none` target.
+While `x86_64-unknown-none` defaults to the `abort` panic strategy, as this
+target does not support the standard library, this default isn't actually
+relevant.
+
+The `std` crate has a `panic_unwind` feature that enables an optional dependency
+on the `panic_unwind` crate.
+
+`core` also has a `panic_immediate_abort` feature which modifies the
+`core::panic!` macro to immediately call the abort intrinsic without calling the
+panic handler. `std` and `alloc` have the same feature which enable the feature
+in `core`. `std`'s feature also adds an immediate abort to its `panic!` macro.
+
+## Cargo
+[background-cargo]: #cargo
+
+Cargo's building of the dependency graph is largely driven by the registry
+index, except for crates from `git` or `path` sources.
+
+[Cargo registries][cargo-docs-registry], like crates.io, are centralised sources
+for crates. A registry's index is the interface between Cargo and the registry
+that Cargo queries to know which crates are available, what their dependencies
```

Comment:
Technically the index can't tell us what crates are available but what crate versions are.

The git index did allow us to query the whole thing but the sparse index does not.  Unless you mean by trial and error...

Reply:
Should be fixed in c740a18